src/result_enumerator.rs

```diff
@@ -96,32 +102,67 @@ impl IWbemClassWrapper {
         }
     }
 
-    pub fn path(&self) -> Result<String, WMIError> {
+    pub fn path(&self) -> WMIResult<String> {
         self.get_property("__Path").and_then(Variant::try_into)
     }
 
-    pub fn class(&self) -> Result<String, WMIError> {
+    pub fn class(&self) -> WMIResult<String> {
         self.get_property("__Class").and_then(Variant::try_into)
     }
 
-    pub fn into_desr<T>(self) -> Result<T, WMIError>
+    pub fn into_desr<T>(self) -> WMIResult<T>
     where
         T: de::DeserializeOwned,
     {
-        from_wbem_class_obj(&self).map_err(WMIError::from)
+        from_wbem_class_obj(self).map_err(WMIError::from)
+    }
+
+    fn serialization_data(&self) -> WMIResult<&'static WbemClassSerializationData> {
+        if self.data.borrow().is_none() {
+            let name = self.class()?;
+            let props = self.list_properties()?;
+            *(self.data.borrow_mut()) = Some(WbemClassSerializationData { name, properties: props });
+        }
+        Ok(unsafe { Box::leak(Box::from_raw(self.data.borrow_mut().as_mut().unwrap() as *mut _)) })
+    }
+}
+
+impl Clone for IWbemClassWrapper {
+    fn clone(&self) -> Self {
+        unsafe { Self::clone(self.inner) }
     }
 }
 
 impl Drop for IWbemClassWrapper {
     fn drop(&mut self) {
         let ptr = self.inner.as_ptr();
-
         unsafe {
             (*ptr).Release();
         }
     }
 }
 
+#[derive(Debug, PartialEq, Eq)]
+struct WbemClassSerializationData {
+    name: String,
+    properties: Vec<String>,
+}
+
+impl Serialize for IWbemClassWrapper {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer
+    {
+        let data = self.serialization_data().map_err(Error::custom)?;
```

Comment:
What you actually want here, which I think avoids all the lifetime problems, is to generate a "map" and not a "struct"
```
impl Serialize for IWbemClassWrapper {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer
    {
        let props = self.list_properties().map_err(Error::custom)?;

        let mut map = serializer.serialize_map(Some(props.len()))?;
        for property in props.iter() {
            let value = self.get_property(&property).unwrap();
            map.serialize_entry(property, &value)?;
        }
        map.end()
    }
}
```

If you could also add to the `src\query.rs::it_works` test to check that it serializes correctly

Reply:
I'm kinda mad this works without any problems. But it is a better way of handling this.

Reply:
😂